### PR TITLE
Aggregate bytecode circuit

### PR DESCRIPTION
**Pull request by DreamWuGit**
Make clean Pr against 0818 branch, will make it ready after all related test passes on new code base!

update: 2022.10.08: 
TODO1: now we use poseidon hash, we have to refactor the hash table in bytecode circuit in zkevm-circuit before re-pick this feature.
TODO2: multi block

### CODE REVIEW

File: zkevm/src/circuit.rs [hunk 1]
```diff
@@ -242,3 +249,74 @@ impl TargetCircuit for PoseidonCircuit {
         hash_rows
     }
 }
+
+pub struct ByteCodeCircuit {}
+impl TargetCircuit for ByteCodeCircuit {
+    type Inner = ByteCodeCircuitImpl<Fr>;
+
+    fn name() -> String {
+        "ByteCode".to_string()
```
Comment (lispc) [inline zkevm/src/circuit.rs:258]: small case to be consistent with others

File: zkevm/src/prover.rs [hunk 2]
```diff
@@ -133,7 +134,7 @@ impl Prover {
     ) {
         let verify_circuit_vk =
             keygen_vk(&self.agg_params, verify_circuit).expect("keygen_vk should not fail");
-
+        println!("come to init_agg_pk_from_verifier_circuit");
```
Comment (lispc) [inline zkevm/src/prover.rs:137]: log::info or log::debug 

"enter Prover::init_agg_pk_from_verifier_circuit"

File: zkevm/src/circuit.rs [hunk 3]
```diff
@@ -242,3 +249,74 @@ impl TargetCircuit for PoseidonCircuit {
...
+                .values()
+                .into_iter()
+                .fold(0usize, |total, v| v.bytes.len() + total)
+        } else {
+            0
+        }
+    }
+    fn get_active_rows(block_result: &BlockResult) -> (Vec<usize>, Vec<usize>) {
+        let witness_block = block_result_to_witness_block::<Fr>(block_result).unwrap();
+        let rows = witness_block
```
Comment (lispc) [inline zkevm/src/circuit.rs:313]: reuse "estimate_rows"?

File: zkevm/src/circuit.rs [hunk 4]
```diff
@@ -242,3 +249,74 @@ impl TargetCircuit for PoseidonCircuit {
...
+
+        // skip setting instance info as current aggr circuit don't support instance column
+        // let num_rows = 1 << *DEGREE;
+        // const NUM_BLINDING_ROWS: usize = 7 - 1;
+        // let instance = vec![vec![r; num_rows - NUM_BLINDING_ROWS]];
+        Ok((inner, vec![]))
+    }
+
+    fn estimate_rows(block_result: &BlockResult) -> usize {
+        if let Ok(witness_block) = block_result_to_witness_block::<Fr>(block_result) {
```
Comment (lispc) [inline zkevm/src/circuit.rs:301]: i suggest `block_result_to_witness_block::<Fr>(block_result).unwrap()`.  block_result_to_witness_block should always success or else it is a bug we should fix

File: zkevm/src/circuit.rs [hunk 5]
```diff
@@ -242,3 +249,74 @@ impl TargetCircuit for PoseidonCircuit {
...
+pub struct ByteCodeCircuit {}
+impl TargetCircuit for ByteCodeCircuit {
+    type Inner = ByteCodeCircuitImpl<Fr>;
+
+    fn name() -> String {
+        "ByteCode".to_string()
+    }
+
+    fn empty() -> Self::Inner {
+        let r = tests::get_randomness();
```
Comment (lispc) [inline zkevm/src/circuit.rs:262]: so the `r` here will be different from block.randomness.  Even it works now, i think let them be same like would be better. Ref: https://github.com/scroll-tech/common-rs/blob/216e372602175f17d8cdbb2e12f443ff268e0c24/zkevm/src/circuit.rs#L118

Comment (DreamWuGit) [reply zkevm/src/circuit.rs:262]: `tests::get_randomness() ` return the `DEFAULT_RAND`  so it pass

Comment (DreamWuGit) [reply zkevm/src/circuit.rs]: update to use `DEFAULT_RAND` directly in [c93e757](https://github.com/scroll-tech/common-rs/pull/61/commits/c93e757d6c57a85dac5a45a693651a5f4d11809c)

File: zkevm/src/circuit.rs (hunk 4 above)
Comment (DreamWuGit) [reply zkevm/src/circuit.rs]: not see the error case, changed to unwrap [c93e757](https://github.com/scroll-tech/common-rs/pull/61/commits/c93e757d6c57a85dac5a45a693651a5f4d11809c)

File: zkevm/src/circuit.rs (hunk 3 above)
Comment (DreamWuGit) [reply zkevm/src/circuit.rs]: update in [c93e757](https://github.com/scroll-tech/common-rs/pull/61/commits/c93e757d6c57a85dac5a45a693651a5f4d11809c) thanks !

File: zkevm/src/prover.rs (hunk 2 above)
Comment (DreamWuGit) [reply zkevm/src/prover.rs]: update in [c93e757](https://github.com/scroll-tech/common-rs/pull/61/commits/c93e757d6c57a85dac5a45a693651a5f4d11809c) thanks for catching the debug info.

File: zkevm/src/circuit.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm/src/circuit.rs]: updated in [c93e757](https://github.com/scroll-tech/common-rs/pull/61/commits/c93e757d6c57a85dac5a45a693651a5f4d11809c)

**Comment by lispc** [comment]
Oh wait you did not agg bytecode circuit at all... You can check the codes of prover you will see many " [0,1,2,3].map" to traverse the target circuits proved results. Your 5th bytecode circuit results is in fact not used at all...

**Comment by lispc** [comment]
I also suggest to squash all commits into 1 and force push

**Comment by DreamWuGit** [comment]
> Oh wait you did not agg bytecode circuit at all... You can check the codes of prover you will see many " [0,1,2,3].map" to traverse the target circuits proved results. Your 5th bytecode circuit results is in fact not used at all...

oh, will check it!

**Comment by DreamWuGit** [comment]
> > Oh wait you did not agg bytecode circuit at all... You can check the codes of prover you will see many " [0,1,2,3].map" to traverse the target circuits proved results. Your 5th bytecode circuit results is in fact not used at all...
> 
> oh, will check it!

fixed in [7d86fe9](https://github.com/scroll-tech/common-rs/pull/61/commits/7d86fe99e2cf7c0586ba20dfa94e52f6191a3a0b) and test `single` and `sushi` trace file, both pass. thanks you for catching this.

**Comment by icemelon** [comment]
@lispc We can close this PR, right?

**Pull request closed without merging**